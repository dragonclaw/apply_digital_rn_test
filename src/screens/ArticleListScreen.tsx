import {StyleSheet, Text, View} from 'react-native';
import React from 'react';

const ArticleListScreen = () => {
  return (
    <View>
      <Text>ArticleScreen</Text>
    </View>
  );
};

export default ArticleListScreen;

const styles = StyleSheet.create({});
