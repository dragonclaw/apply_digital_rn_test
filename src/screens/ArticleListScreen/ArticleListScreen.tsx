import {Text, View} from 'react-native';
import React, {useEffect, useState} from 'react';
import articles from '../../services/articles/articles';
import ArticleList from '../../components/ArticleListComponent/ArticleList';
import AsyncStorage from '@react-native-async-storage/async-storage';
import NetInfo from '@react-native-community/netinfo';
import notifee, {EventType} from '@notifee/react-native';
import {NavigationProp, useNavigation} from '@react-navigation/native';
import {NavigationTypes} from '../../components/CardArticleComponent/CardArticleComponent.types';
import BackgroundFetch from 'react-native-background-fetch';
import LoadingComponent from '../../components/LoadingComponent/LoadingComponent';
import {compareArrays} from '../../utils/compareArrays';

const ArticleListScreen = () => {
  const navigation = useNavigation<NavigationProp<NavigationTypes>>();
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setError] = useState(false);
  //TODO: need to refactor to use context
  const [data, setData] = useState([]);
  const [isOffline, setIsOffline] = useState(false);

  //TODO: need to refactor to only fetch and save into context
  const fetchList = async () => {
    try {
      setIsLoading(true);
      const articlesData = await articles.getArticles();
      const deletedArticles = JSON.parse(
        (await AsyncStorage.getItem('deletedArticles')) as string,
      );
      const filteredArticles = articlesData.filter(
        (article: {story_id: any}) => {
          return !deletedArticles?.includes(article.story_id);
        },
      );
      setData(filteredArticles);
      await AsyncStorage.setItem('articles', JSON.stringify(filteredArticles));
      setIsLoading(false);
    } catch (err) {
      setIsLoading(false);
      setError(true);
    }
  };

  const loadFromStorage = async () => {
    try {
      const storedData = await AsyncStorage.getItem('articles');
      if (storedData) {
        setData(JSON.parse(storedData));
      }
    } catch (err) {
      setError(true);
    }
  };

  useEffect(() => {
    const checkConnectivity = async () => {
      const state = await NetInfo.fetch();
      setIsOffline(!state.isConnected);
    };
    checkConnectivity();
    if (!isOffline) {
      fetchList();
    } else {
      loadFromStorage();
    }
  }, [isOffline]);

  useEffect(() => {
    initBackgroundFetch();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const initBackgroundFetch = async () => {
    // BackgroundFetch event handler.
    const onEvent = async (taskId: string | undefined) => {
      console.log('[BackgroundFetch] task: ', taskId);
      console.log('THIS SHOULD BE EXECUTED IN THE BACKGROUND');
      // Do your background work...
      const currentArticles = data;
      const articlesData = await articles.getArticles();
      if (!compareArrays(currentArticles, articlesData)) {
        // Create a channel (required for Android)
        const channelId = await notifee.createChannel({
          id: 'default',
          name: 'Default Channel',
        });

        // Display a notification
        await notifee.displayNotification({
          title: articlesData[0].title,
          body: articlesData[0].author,
          android: {
            channelId,

            // pressAction is needed if you want the notification to open the app when pressed
            pressAction: {
              id: articlesData[0].url,
              launchActivity: 'default',
            },
          },
        });
      }

      // IMPORTANT:  You must signal to the OS that your task is complete.
      BackgroundFetch.finish(taskId);
    };

    // Timeout callback is executed when your Task has exceeded its allowed running-time.
    // You must stop what you're doing immediately BackgroundFetch.finish(taskId)
    const onTimeout = async (taskId: string | undefined) => {
      console.warn('[BackgroundFetch] TIMEOUT task: ', taskId);
      BackgroundFetch.finish(taskId);
    };

    // Initialize BackgroundFetch only once when component mounts.
    let status = await BackgroundFetch.configure(
      {minimumFetchInterval: 15},
      onEvent,
      onTimeout,
    );

    console.log('[BackgroundFetch] configure status: ', status);
  };

  notifee.onForegroundEvent(({type, detail}) => {
    console.log('FOREGROUND');
    console.log('type', type);
    console.log('detail', detail);
    console.log(EventType);
    if (
      type === EventType.PRESS &&
      detail.pressAction &&
      detail.pressAction.id
    ) {
      console.log('NAVIGATION IN BACKGROUND');
      navigation.navigate('SingleArticleScreen', {
        url: detail.pressAction.id,
      });
      console.log(
        'User pressed an action with the id ON THE BACKGROUND: ',
        detail.pressAction.id,
      );
    }
  });

  notifee.onBackgroundEvent(async ({type, detail}) => {
    console.log('BACKGROUND');
    console.log('type', type);
    console.log('detail', detail);
    console.log(EventType);
    if (
      type === EventType.PRESS &&
      detail.pressAction &&
      detail.pressAction.id
    ) {
      console.log('NAVIGATION IN BACKGROUND');
      navigation.navigate('SingleArticleScreen', {
        url: detail.pressAction.id,
      });
      console.log(
        'User pressed an action with the id ON THE BACKGROUND: ',
        detail.pressAction.id,
      );
    }
  });

  return isLoading ? (
    <LoadingComponent />
  ) : isError && data.length === 0 ? (
    // TODO: change for a error component
    <View>
      <Text>Error fetching the data</Text>
    </View>
  ) : (
    <>
      {isOffline && (
        <View>
          <Text>is offline</Text>
        </View>
      )}

      <ArticleList data={data} onRefresh={fetchList} />
    </>
  );
};

export default ArticleListScreen;
