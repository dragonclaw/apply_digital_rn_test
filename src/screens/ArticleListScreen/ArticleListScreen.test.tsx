import React from 'react';
import {render} from '@testing-library/react-native';
import ArticleListScreen from './ArticleListScreen'; // Adjust the import path as necessary
import AsyncStorage from '@react-native-async-storage/async-storage';

// Mock the articles service and notifee
jest.mock('../../services/articles/articles');
// jest.mock('@notifee/react-native');

// Mock AsyncStorage
jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
}));

// Mock notifee
jest.mock('@notifee/react-native', () => ({
  createChannel: jest.fn().mockResolvedValue('default'),
  onForegroundEvent: jest.fn(),
  onBackgroundEvent: jest.fn(),
  displayNotification: jest.fn(),
}));

jest.mock('@react-native-community/netinfo', () => ({
  fetch: jest.fn().mockResolvedValue({
    isConnected: true,
  }),
}));

jest.mock('react-native-background-fetch', () => ({
  configure: jest.fn(),
  scheduleTask: jest.fn(),
}));

jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(),
}));

describe('ArticleListScreen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches the snapshot', async () => {
    // Mock AsyncStorage response
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(
      JSON.stringify([{title: 'Test Article', author: 'Test Author'}]),
    );

    const {toJSON} = render(<ArticleListScreen />);
    expect(toJSON()).toMatchSnapshot();
  });
});
