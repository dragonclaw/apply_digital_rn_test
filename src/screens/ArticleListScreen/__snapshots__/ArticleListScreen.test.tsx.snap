// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleListScreen matches the snapshot 1`] = `
<View
  className="sc-gnpbhQ ftrdcH"
>
  <ActivityIndicator
    className="sc-la-DxNn hhtHDd"
    color="#aaa"
    size={50}
  />
</View>
`;
